File: packages/sandboxd/src/pi_proxy/idempotency.rs
use std::sync::{Arc, Mutex};

use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::idempotency::store::{IdempotencyStore, IdempotencyStoreError};
use crate::idempotency::{
    AgentRuntimeId, CompleteIdempotencyOperation, IdempotencyOperation, IdempotencyRecord,
    IdempotencyRecordStatus, RepeatedRequestOutcome, RequestFingerprint, StartIdempotencyOperation,
};
use crate::pi_proxy::PiProxyError;
use crate::time::{Clock, SystemClock, format_rfc3339_timestamp};

pub(super) type SharedIdempotencyStore = Arc<Mutex<IdempotencyStore>>;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(super) struct PiProxyIdempotency {
    key: String,
    operation: PiProxyIdempotencyOperation,
    request_fingerprint: RequestFingerprint,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
enum PiProxyIdempotencyOperation {
    CreateConversation,
    SubmitPayload,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(super) struct StoredPiProxyResponse {
    pub(super) payload: Value,
}

#[derive(Debug, Clone)]
pub(super) struct StartedPiOperation {
    idempotency: PiProxyIdempotency,
    operation: IdempotencyOperation,
    method: String,
}

impl StartedPiOperation {
    pub(super) fn method(&self) -> &str {
        &self.method
    }
}

pub(super) enum PiIdempotencyAction {
    Disabled,
    Forward(StartedPiOperation),
    Replay(StoredPiProxyResponse),
    Reject { message: String },
}

pub(super) fn prepare_pi_idempotency(
    idempotency: Option<&PiProxyIdempotency>,
    method: &str,
    store: Option<&SharedIdempotencyStore>,
) -> PiIdempotencyAction {
    let Some(idempotency) = idempotency.cloned() else {
        return PiIdempotencyAction::Disabled;
    };
    let operation = match (idempotency.operation, method) {
        (PiProxyIdempotencyOperation::CreateConversation, "pi/createConversation") => {
            IdempotencyOperation::CreateConversation
        }
        (PiProxyIdempotencyOperation::SubmitPayload, "pi/prompt" | "pi/steer" | "pi/followUp") => {
            IdempotencyOperation::SubmitPayload
        }
        (PiProxyIdempotencyOperation::CreateConversation, _) => {
            return PiIdempotencyAction::Reject {
                message: "Pi createConversation idempotency requires pi/createConversation."
                    .to_string(),
            };
        }
        (PiProxyIdempotencyOperation::SubmitPayload, _) => {
            return PiIdempotencyAction::Reject {
                message:
                    "Pi submitPayload idempotency requires pi/prompt, pi/steer, or pi/followUp."
                        .to_string(),
            };
        }
    };
    let Some(store) = store else {
        return PiIdempotencyAction::Reject {
            message: "Pi idempotency store is not configured.".to_string(),
        };
    };
    let now = match now_timestamp() {
        Ok(now) => now,
        Err(error) => {
            return PiIdempotencyAction::Reject {
                message: error.to_string(),
            };
        }
    };
    let start_result = match lock_store(store).and_then(|mut store| {
        match store.get_by_key(AgentRuntimeId::Pi, operation.clone(), &idempotency.key) {
            Ok(record) => Ok(PiStartResult::Existing(record.clone())),
            Err(IdempotencyStoreError::MissingRecord { .. }) => store
                .start_operation(StartIdempotencyOperation {
                    key: idempotency.key.clone(),
                    runtime_id: AgentRuntimeId::Pi,
                    operation: operation.clone(),
                    request_fingerprint: idempotency.request_fingerprint.clone(),
                    now,
                })
                .map(PiStartResult::Created)
                .map_err(map_store_error),
            Err(error) => Err(map_store_error(error)),
        }
    }) {
        Ok(result) => result,
        Err(error) => {
            return PiIdempotencyAction::Reject {
                message: error.to_string(),
            };
        }
    };
    match start_result {
        PiStartResult::Created(record) => {
            if record.status != IdempotencyRecordStatus::Started {
                return PiIdempotencyAction::Reject {
                    message: format!(
                        "Pi idempotency key '{}' did not start in started status.",
                        record.key
                    ),
                };
            }
            PiIdempotencyAction::Forward(StartedPiOperation {
                idempotency,
                operation,
                method: method.to_string(),
            })
        }
        PiStartResult::Existing(record) => {
            let outcome = match record.classify_repeated_request(&idempotency.request_fingerprint) {
                Ok(outcome) => outcome,
                Err(error) => {
                    return PiIdempotencyAction::Reject {
                        message: error.to_string(),
                    };
                }
            };
            if outcome == RepeatedRequestOutcome::Completed {
                return match record.response {
                    Some(response) => {
                        match serde_json::from_value::<StoredPiProxyResponse>(response) {
                            Ok(response) => PiIdempotencyAction::Replay(response),
                            Err(error) => PiIdempotencyAction::Reject {
                                message: format!("Pi idempotency response is invalid: {error}"),
                            },
                        }
                    }
                    None => PiIdempotencyAction::Reject {
                        message: format!(
                            "Pi idempotency key '{}' completed without a response.",
                            record.key
                        ),
                    },
                };
            }
            PiIdempotencyAction::Reject {
                message: format!(
                    "Pi idempotency key '{}' has unresolved status {:?}.",
                    record.key, record.status
                ),
            }
        }
    }
}

enum PiStartResult {
    Created(IdempotencyRecord),
    Existing(IdempotencyRecord),
}

pub(super) fn complete_pi_idempotency(
    store: &SharedIdempotencyStore,
    started: StartedPiOperation,
    response: StoredPiProxyResponse,
    provider_conversation_id: Option<String>,
) -> Result<(), PiProxyError> {
    let now = now_timestamp()?;
    let response_value = serde_json::to_value(response)
        .map_err(|error| PiProxyError::InvalidRequest(error.to_string()))?;
    let mut store = lock_store(store)?;
    store
        .mark_completed(
            AgentRuntimeId::Pi,
            started.operation,
            &started.idempotency.key,
            CompleteIdempotencyOperation {
                request_fingerprint: started.idempotency.request_fingerprint,
                provider_conversation_id,
                provider_execution_id: None,
                runtime_artifact_hint: Some(json!({ "method": started.method })),
                response: response_value,
                now,
            },
        )
        .map(|_| ())
        .map_err(map_store_error)
}

fn lock_store(
    store: &SharedIdempotencyStore,
) -> Result<std::sync::MutexGuard<'_, IdempotencyStore>, PiProxyError> {
    store.lock().map_err(|error| {
        PiProxyError::InvalidRequest(format!("Pi idempotency store lock is poisoned: {error}"))
    })
}

fn map_store_error(error: IdempotencyStoreError) -> PiProxyError {
    PiProxyError::InvalidRequest(error.to_string())
}

fn now_timestamp() -> Result<String, PiProxyError> {
    format_rfc3339_timestamp(SystemClock.now_system_time())
        .map_err(|error| PiProxyError::InvalidRequest(error.to_string()))
}

File: packages/sandboxd/piproxy/idempotency.go
package piproxy

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/mistle/sandboxd/idempotency"
)

type SharedIdempotencyStore struct {
	mutex sync.Mutex
	store *idempotency.Store
}

type Idempotency struct {
	Key                string                         `json:"key"`
	Operation          IdempotencyOperation           `json:"operation"`
	RequestFingerprint idempotency.RequestFingerprint `json:"requestFingerprint"`
}

type IdempotencyOperation string

const (
	IdempotencyOperationCreateConversation IdempotencyOperation = "createConversation"
	IdempotencyOperationSubmitPayload      IdempotencyOperation = "submitPayload"
)

type StoredResponse struct {
	Payload map[string]any `json:"payload"`
}

type startedOperation struct {
	idempotency Idempotency
	operation   idempotency.IdempotencyOperation
	method      string
}

type idempotencyActionKind string

const (
	idempotencyActionDisabled idempotencyActionKind = "disabled"
	idempotencyActionForward  idempotencyActionKind = "forward"
	idempotencyActionReplay   idempotencyActionKind = "replay"
	idempotencyActionReject   idempotencyActionKind = "reject"
)

type idempotencyAction struct {
	kind    idempotencyActionKind
	started *startedOperation
	replay  *StoredResponse
	message string
}

func NewSharedIdempotencyStore(store *idempotency.Store) *SharedIdempotencyStore {
	return &SharedIdempotencyStore{store: store}
}

func prepareIdempotency(request jsonRPCRequest, store *SharedIdempotencyStore) idempotencyAction {
	if len(request.Idempotency) == 0 || bytes.Equal(bytes.TrimSpace(request.Idempotency), []byte("null")) {
		return idempotencyAction{kind: idempotencyActionDisabled}
	}
	var requestIdempotency Idempotency
	if err := decodeStrictPiIdempotency(request.Idempotency, &requestIdempotency); err != nil {
		return idempotencyAction{kind: idempotencyActionReject, message: fmt.Sprintf("Pi idempotency envelope is invalid: %v", err)}
	}
	operation, rejectMessage := idempotencyOperationForPiMethod(requestIdempotency.Operation, request.Method)
	if rejectMessage != "" {
		return idempotencyAction{kind: idempotencyActionReject, message: rejectMessage}
	}
	if store == nil {
		return idempotencyAction{kind: idempotencyActionReject, message: "Pi idempotency store is not configured."}
	}
	store.mutex.Lock()
	defer store.mutex.Unlock()

	record, err := store.store.GetByKey(idempotency.AgentRuntimePi, operation, requestIdempotency.Key)
	if err != nil {
		var missing idempotency.MissingRecordError
		if !errors.As(err, &missing) {
			return idempotencyAction{kind: idempotencyActionReject, message: err.Error()}
		}
		record, err = store.store.StartOperation(idempotency.StartOperation{
			Key:                requestIdempotency.Key,
			RuntimeID:          idempotency.AgentRuntimePi,
			Operation:          operation,
			RequestFingerprint: requestIdempotency.RequestFingerprint,
			Now:                nowTimestamp(),
		})
		if err != nil {
			return idempotencyAction{kind: idempotencyActionReject, message: err.Error()}
		}
		if record.Status != idempotency.IdempotencyRecordStarted {
			return idempotencyAction{kind: idempotencyActionReject, message: fmt.Sprintf("Pi idempotency key %q did not start in started status.", record.Key)}
		}
		return idempotencyAction{kind: idempotencyActionForward, started: &startedOperation{
			idempotency: requestIdempotency,
			operation:   operation,
			method:      request.Method,
		}}
	}

	outcome, err := record.ClassifyRepeatedRequest(requestIdempotency.RequestFingerprint)
	if err != nil {
		return idempotencyAction{kind: idempotencyActionReject, message: err.Error()}
	}
	if outcome != idempotency.RepeatedRequestCompleted {
		return idempotencyAction{kind: idempotencyActionReject, message: fmt.Sprintf("Pi idempotency key %q has unresolved status %s.", record.Key, record.Status)}
	}
	response, err := storedResponseFromRecord(record)
	if err != nil {
		return idempotencyAction{kind: idempotencyActionReject, message: fmt.Sprintf("Pi idempotency response is invalid: %v", err)}
	}
	return idempotencyAction{kind: idempotencyActionReplay, replay: &response}
}

func completeIdempotency(store *SharedIdempotencyStore, started startedOperation, response StoredResponse) error {
	providerConversationID := providerConversationIDFromResponse(started.method, response.Payload)
	store.mutex.Lock()
	defer store.mutex.Unlock()
	_, err := store.store.MarkCompleted(idempotency.AgentRuntimePi, started.operation, started.idempotency.Key, idempotency.CompleteOperation{
		RequestFingerprint:     started.idempotency.RequestFingerprint,
		ProviderConversationID: providerConversationID,
		ProviderExecutionID:    nil,
		RuntimeArtifactHint:    map[string]any{"method": started.method},
		Response:               response,
		Now:                    nowTimestamp(),
	})
	return err
}

func deleteStartedIdempotency(store *SharedIdempotencyStore, started startedOperation) error {
	store.mutex.Lock()
	defer store.mutex.Unlock()
	return store.store.DeleteStarted(
		idempotency.AgentRuntimePi,
		started.operation,
		started.idempotency.Key,
		started.idempotency.RequestFingerprint,
	)
}

func idempotencyOperationForPiMethod(operation IdempotencyOperation, method string) (idempotency.IdempotencyOperation, string) {
	switch operation {
	case IdempotencyOperationCreateConversation:
		if method != "pi/createConversation" {
			return "", "Pi createConversation idempotency requires pi/createConversation."
		}
		return idempotency.IdempotencyOperationCreateConversation, ""
	case IdempotencyOperationSubmitPayload:
		if method != "pi/prompt" && method != "pi/steer" && method != "pi/followUp" {
			return "", "Pi submitPayload idempotency requires pi/prompt, pi/steer, or pi/followUp."
		}
		return idempotency.IdempotencyOperationSubmitPayload, ""
	default:
		return "", "Pi idempotency operation is not supported."
	}
}

func providerConversationIDFromResponse(method string, payload map[string]any) *string {
	result, ok := payload["result"].(map[string]any)
	if !ok {
		return nil
	}
	switch method {
	case "pi/createConversation":
		if providerConversationID, ok := result["providerConversationId"].(string); ok {
			return &providerConversationID
		}
	case "pi/prompt", "pi/steer", "pi/followUp":
		if sessionFile, ok := result["sessionFile"].(string); ok {
			return &sessionFile
		}
	}
	return nil
}

func storedResponseFromRecord(record idempotency.IdempotencyRecord) (StoredResponse, error) {
	serialized, err := json.Marshal(record.Response)
	if err != nil {
		return StoredResponse{}, err
	}
	var response StoredResponse
	if err := json.Unmarshal(serialized, &response); err != nil {
		return StoredResponse{}, err
	}
	return response, nil
}

func nowTimestamp() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}
